### PR TITLE
Add Facebook event logging functionality

### PR DESCRIPTION
**Pull request by fbcouch**
This commit introduces comprehensive event logging capabilities to track user actions and conversions in Facebook Analytics.

Key changes:
- Added logEvent() method for custom events with parameters
- Added logPurchaseEvent() convenience method for purchase tracking
- Added logCompleteRegistration() method for registration tracking
- Implemented Android support using AppEventsLogger from Facebook SDK
- Implemented iOS support using AppEvents.shared.logEvent from FBSDKCoreKit
- Enhanced example app with event logging demonstrations
- Updated CHANGELOG for version 1.3.0

All events require prior SDK initialization via consentProvided() for privacy compliance.

### CODE REVIEW

**Comment by twinsundev** [comment]
### 🤖 Scarif Code Review

## Summary

This PR adds comprehensive Facebook Analytics event logging capabilities to the plugin, introducing three new public methods (`logEvent()`, `logPurchaseEvent()`, and `logCompleteRegistration()`). The implementation is well-structured with full platform support for both iOS and Android, excellent documentation, and good integration with the existing codebase.

**Overall Assessment**: The changes are well-implemented with clear documentation. However, there are several areas that need attention before merging, particularly around testing, validation, and privacy compliance.

---

## Priority Issues

### Critical

**1. Missing Unit Tests**
- **Issue**: No tests were added for the new event logging functionality
- **Impact**: Changes to a public API without test coverage increases risk of regressions
- **Recommendation**: Add unit tests covering:
  - Valid event logging with/without parameters
  - Invalid event names (empty, null, special characters, exceeding 40 character limit)
  - Parameter type conversion edge cases
  - Error handling scenarios
  - SDK initialization requirements

### High Priority

**2. Input Validation Missing**
- **Issue**: The Dart layer doesn't validate input before sending to native platforms
  - Event names can exceed Facebook's 40-character limit
  - No validation for currency codes (should be 3-letter ISO 4217)
  - No validation for numeric values (e.g., negative amounts in purchases)
- **Location**: `lib/flutter_facebook_app_links.dart:121-172`
- **Recommendation**: Add validation in the Dart layer:
  ```dart
  static Future<void> logEvent(String eventName, [Map<String, dynamic>? parameters]) async {
    if (eventName.isEmpty || eventName.length > 40) {
      throw ArgumentError('Event name must be 1-40 characters');
    }
    if (\!RegExp(r'^[a-zA-Z0-9_]+$').hasMatch(eventName)) {
      throw ArgumentError('Event name must be alphanumeric with underscores only');
    }
    // ... rest of implementation
  }
  ```

**3. Potential Privacy Compliance Issue**
- **Issue**: Events can be logged even if `consentProvided()` was never called or if `consentRevoked()` was called
- **Impact**: Could violate GDPR/privacy regulations that are explicitly mentioned in the README
- **Location**: Android `FlutterFacebookAppLinksPlugin.java:185-234`, iOS `SwiftFlutterFacebookAppLinksPlugin.swift:107-151`
- **Recommendation**: Track consent state and prevent event logging without consent, or document this responsibility clearly

---

## Suggestions

### Code Quality

**1. Type Conversion Redundancy in iOS** (`ios/Classes/SwiftFlutterFacebookAppLinksPlugin.swift:131-144`)
- The type checking has redundancy: `NSNumber`, `Int`, `Double`, and `Bool` checks overlap since `Int`, `Double`, and `Bool` can all match `NSNumber`
- Consider reordering checks to test specific types before `NSNumber`:
  ```swift
  if let intValue = value as? Int {
    eventParameters[parameterName] = intValue
  } else if let doubleValue = value as? Double {
    eventParameters[parameterName] = doubleValue
  } else if let boolValue = value as? Bool {
    eventParameters[parameterName] = boolValue
  } else if let stringValue = value as? String {
    eventParameters[parameterName] = stringValue
  } else {
    eventParameters[parameterName] = String(describing: value)
  }
  ```

**2. Error Handling Inconsistency**
- `logEvent()` rethrows exceptions, while `activateSDK()` silently catches and logs
- **Recommendation**: Maintain consistent error handling strategy across the API

**3. Potential Null Pointer Exception** (`android/src/main/java/it/remedia/flutter_facebook_app_links/FlutterFacebookAppLinksPlugin.java:201`)
- `AppEventsLogger.newLogger(mContext)` could fail if `mContext` is null
- **Recommendation**: Add null check:
  ```java
  if (mContext == null) {
    result.error("CONTEXT_NULL", "Plugin context is not initialized", null);
    return;
  }
  ```

**4. Convenience Method Parameter Mutation**
- `logPurchaseEvent()` mutates the passed parameters map
- **Recommendation**: Already handled correctly with `Map.from()` - good defensive programming\!

### Documentation

**5. Missing README Documentation**
- The README doesn't document the new event logging features
- **Recommendation**: Add a section documenting:
  - How to log events
  - When to log events (after consent)
  - Example code snippets
  - Link to Facebook Analytics event reference

**6. Missing Parameter Documentation**
- Standard Facebook parameter names like `fb_content_id`, `fb_num_items` should be documented or referenced
- Consider creating constants for standard parameter keys to prevent typos

### Testing

**7. Example App Could Show Error Scenarios**
- The example only demonstrates successful event logging
- **Recommendation**: Add examples showing:
  - What happens when SDK isn't initialized
  - Error handling patterns
  - Best practices for production use

---

## Security Assessment

**No Critical Security Issues Identified**

✅ **Good Practices Observed:**
- Input is properly sanitized through type checking before native calls
- SDK initialization checks are in place
- No SQL injection, XSS, or similar vulnerabilities (not applicable for this code)

⚠️ **Privacy Considerations:**
- Ensure users cannot log PII (Personally Identifiable Information) without proper consent
- Consider documenting privacy best practices for event parameters
- Facebook SDK will collect device data - ensure this is disclosed in privacy policy

---

## Positive Highlights

🎉 **Well Done:**
- **Excellent Documentation**: Comprehensive inline documentation with examples
- **Cross-Platform Consistency**: Both iOS and Android implementations mirror each other
- **Backward Compatibility**: No breaking changes to existing API
- **Proper Error Handling**: Platform errors are caught and reported appropriately
- **CHANGELOG Updated**: Version bumped and changes documented clearly
- **Example Integration**: Practical examples in the demo app help developers understand usage
- **Type Safety**: Proper type conversion handling for parameters on both platforms

---

## Additional Notes

**Recommendations for Future Work:**
1. Consider adding a method to flush events immediately (for testing)
2. Consider adding support for standard Facebook event constants as Dart enums
3. Consider batch event logging for performance optimization
4. Document event deduplication strategy if users need it

**Testing Suggestions:**
- Test with Facebook Events Manager to verify events are received
- Test offline scenarios and event queuing behavior
- Test with various parameter combinations and edge cases

---

## Conclusion

This is a **solid implementation** that adds valuable functionality to the plugin. The code is clean, well-documented, and follows existing patterns. The main concerns are around **testing coverage** and **input validation**, which should be addressed before merging to production.

**Recommended Actions Before Merge:**
1. Add comprehensive unit tests
2. Add input validation in Dart layer
3. Document consent requirements more explicitly
4. Update README with new functionality

Great work on maintaining code quality and documentation\! 🚀

File: lib/flutter_facebook_app_links.dart [hunk 1]
```diff
@@ -76,4 +76,130 @@ class FlutterFacebookAppLinks {
...
+  /// await FlutterFacebookAppLinks.logEvent('product_viewed', {
+  ///   'fb_content_id': 'product_123',
+  ///   'fb_content_type': 'product',
+  ///   '_valueToSum': 29.99,
+  ///   'fb_currency': 'USD'
+  /// });
+  /// ```
+  ///
+  /// Throws [PlatformException] if the platform-specific call fails
+  static Future<void> logEvent(String eventName, [Map<String, dynamic>? parameters]) async {
```
Comment (twinsundev) [inline lib/flutter_facebook_app_links.dart:121]: **Input Validation Needed**

The Dart layer should validate event names before sending to native platforms. Facebook has specific requirements:
- Maximum 40 characters
- Alphanumeric characters and underscores only

Consider adding validation:
```dart
if (eventName.isEmpty || eventName.length > 40) {
  throw ArgumentError('Event name must be 1-40 characters');
}
if (\!RegExp(r'^[a-zA-Z0-9_]+$').hasMatch(eventName)) {
  throw ArgumentError('Event name must be alphanumeric with underscores only');
}
```

This prevents invalid events from being sent and provides better error messages to developers.

File: android/src/main/java/it/remedia/flutter_facebook_app_links/FlutterFacebookAppLinksPlugin.java [hunk 2]
```diff
@@ -176,4 +181,56 @@ public void run() {
...
+      String eventName = call.argument("eventName");
+      Map<String, Object> parameters = call.argument("parameters");
+
+      if (eventName == null || eventName.isEmpty()) {
+        result.error("INVALID_ARGUMENTS", "Event name cannot be null or empty", null);
+        return;
+      }
+
+      // Create AppEventsLogger instance
+      AppEventsLogger logger = AppEventsLogger.newLogger(mContext);
```
Comment (twinsundev) [inline android/src/main/java/it/remedia/flutter_facebook_app_links/FlutterFacebookAppLinksPlugin.java:201]: **Potential Null Context Issue**

Consider adding a null check for `mContext` before using it:

```java
if (mContext == null) {
  result.error("CONTEXT_NULL", "Plugin context is not initialized", null);
  return;
}
```

While unlikely in normal operation, this defensive check prevents potential crashes and provides clearer error messages if the plugin is used before proper initialization.

File: lib/flutter_facebook_app_links.dart [hunk 3]
```diff
@@ -76,4 +76,130 @@ class FlutterFacebookAppLinks {
...
+  /// ```dart
+  /// await FlutterFacebookAppLinks.logPurchaseEvent(
+  ///   49.99,
+  ///   'USD',
+  ///   {'fb_content_id': 'sku_12345', 'fb_num_items': 2}
+  /// );
+  /// ```
+  ///
+  /// Throws [PlatformException] if the platform-specific call fails
+  static Future<void> logPurchaseEvent(
```
Comment (twinsundev) [inline lib/flutter_facebook_app_links.dart:154]: **Currency Code Validation Missing**

The purchase event should validate the currency code format:

```dart
if (currency.length != 3 || !RegExp(r'^[A-Z]+$').hasMatch(currency)) {
  throw ArgumentError('Currency must be a 3-letter ISO 4217 code (e.g., USD, EUR, GBP)');
}
if (amount < 0) {
  throw ArgumentError('Purchase amount cannot be negative');
}
```

This ensures compliance with Facebook's expected format and prevents data quality issues in analytics.

File: ios/Classes/SwiftFlutterFacebookAppLinksPlugin.swift [hunk 4]
```diff
@@ -101,4 +103,50 @@ public class SwiftFlutterFacebookAppLinksPlugin: NSObject, FlutterPlugin {
...
+
+    // Convert parameters to proper types for AppEvents
+    var eventParameters: [AppEvents.ParameterName: Any] = [:]
+
+    for (key, value) in parameters {
+      // Create custom parameter name from the key string
+      let parameterName = AppEvents.ParameterName(rawValue: key)
+
+      // Add the value with proper type handling
+      if let stringValue = value as? String {
```
Comment (twinsundev) [inline ios/Classes/SwiftFlutterFacebookAppLinksPlugin.swift:131]: **Type Checking Order Optimization**

The current type checks have some redundancy since `Int`, `Double`, and `Bool` can all match as `NSNumber` in Swift. Consider removing the `NSNumber` check or reordering:

```swift
// Check specific types first
if let stringValue = value as? String {
  eventParameters[parameterName] = stringValue
} else if let intValue = value as? Int {
  eventParameters[parameterName] = intValue
} else if let doubleValue = value as? Double {
  eventParameters[parameterName] = doubleValue
} else if let boolValue = value as? Bool {
  eventParameters[parameterName] = boolValue
} else {
  eventParameters[parameterName] = String(describing: value)
}
```

This ensures more predictable type handling and avoids the NSNumber overlap.

**Comment by sarahemaginnis** [comment]
You can close this pull request - I implemented the logging and added in the suggestions from the TwinSunDev bot into my own branch/pre-existing pull request.
